### PR TITLE
[IMP] stock: Zero Quantity Count

### PR DESCRIPTION
**Pull request by arbaes**
WIP

### CODE REVIEW

File: addons/stock/models/stock_move.py [hunk 1]
```diff
@@ -1150,8 +1150,10 @@ def _action_done(self, cancel_backorder=False):
         if self.env.context.get('is_scrap'):
             return moves_todo
 
-        if picking and not cancel_backorder:
-            picking._create_backorder()
+        if pickings:
```
Comment (Whenrow) [inline addons/stock/models/stock_move.py]: _check_backorder() return true or false so i would do
```python 
if pickings._check_backorder() and not cancel_backorder:
    pickings._create_backorder
```

File: addons/stock/models/stock_move.py [hunk 2]
```diff
@@ -1141,7 +1141,7 @@ def _action_done(self, cancel_backorder=False):
                 .filtered(lambda p: p.quant_ids and len(p.quant_ids) > 1):
             if len(result_package.quant_ids.mapped('location_id')) > 1:
                 raise UserError(_('You cannot move the same package content more than once in the same transfer or split the same package into two location.'))
-        picking = moves_todo and moves_todo[0].picking_id or False
+        pickings = moves_todo and moves_todo.mapped('picking_id') or False
```
Comment (fmdl) [inline addons/stock/models/stock_move.py]: or more simple : `pickings = moves_todo.mapped('picking_id')`

**Comment by sle-odoo** [comment]
commit message

File: addons/stock/models/stock_move.py [hunk 3]
```diff
@@ -1214,8 +1214,10 @@ def _action_done(self, cancel_backorder=False):
         if self.env.context.get('is_scrap'):
             return moves_todo
 
-        if picking and not cancel_backorder:
-            picking._create_backorder()
+        if pickings:
+            pickings_to_backorder = pickings._check_backorder()
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: change of behavior

File: addons/stock/models/stock_picking.py [hunk 4]
```diff
@@ -747,40 +746,35 @@ def button_validate(self):
...
-            }
+    def _finalize_validation(self):
+        """ Raise wizards if needed before and after applying action_done() on the picking(s) """
+
+        if self._check_no_quantities_done() and not self._context.get('skip_no_quantity_check'):
+            return self.action_generate_immediate_transfer_wizard()
 
         # Check backorder should check for other barcodes
-        if self._check_backorder():
+        if self._check_backorder() and not self._context.get('skip_backorder_check'):
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: you could simply check if cancel backorder is in the context?

File: addons/stock/models/stock_picking.py [hunk 5]
```diff
@@ -747,40 +746,35 @@ def button_validate(self):
...
             return self.action_generate_backorder_wizard()
         self.action_done()
-        return
+        return True
+
+    def action_generate_immediate_transfer_wizard(self):
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: why an action why public

File: addons/stock/wizard/stock_backorder_confirmation.py [hunk 6]
```diff
@@ -14,16 +14,17 @@ class StockBackorderConfirmation(models.TransientModel):
...
                 for move in pick_id.move_lines:
                     if float_compare(move.product_uom_qty, move.quantity_done, precision_rounding=move.product_uom.rounding) > 0:
                         moves_to_log[move] = (move.quantity_done, move.product_uom_qty)
                 pick_id._log_less_quantities_than_expected(moves_to_log)
-        self.pick_ids.with_context({'cancel_backorder': cancel_backorder}).action_done()
+        return self.pick_ids.with_context({'cancel_backorder': cancel_backorder, 'skip_backorder_check': True})._finalize_validation()
 
     def process(self):
-        self._process()
+        return self._process()
```
Comment (sle-odoo) [inline addons/stock/wizard/stock_backorder_confirmation.py]: cancel_backorder=False

File: addons/stock_picking_batch/models/stock_picking_batch.py [hunk 7]
```diff
@@ -64,44 +64,7 @@ def done(self):
...
-                'views': [(view.id, 'form')],
-                'view_id': view.id,
-                'target': 'new',
-                'res_id': wiz.id,
-                'context': self.env.context,
-            }
-        if picking_to_backorder:
-            return picking_to_backorder.action_generate_backorder_wizard()
-        return True
+        return pickings._finalize_validation()
```
Comment (sle-odoo) [inline addons/stock_picking_batch/models/stock_picking_batch.py]: pickings.button_validate

File: addons/stock/models/stock_picking.py [hunk 8]
```diff
@@ -802,6 +846,15 @@ def action_toggle_is_locked(self):
         self.is_locked = not self.is_locked
         return True
 
+    def _check_no_quantities_done(self):
+        picking_no_quantities_done = self.env['stock.picking']
+        precision_digits = self.env['decimal.precision'].precision_get('Product Unit of Measure')
```
Comment (amoyaux) [inline addons/stock/models/stock_picking.py]: sudo

**Pull request closed without merging**